File: bioinformatics/deconv.r
# color compensation / multi-channel deconvolution


# function: get cross-over constant matrix k
get_k <- function(
    db_conn, # MySQL database connection
    dcv_exp_info, # deconvolution experiment ids, i.e. the data in these experiments are for constructing deconvolution matrix. a named vector (by channel) where each element is the experiment id using the target dye of each channel
    well_proc='dim3' # options: 'mean', 'dim3'.
    ) { 
    
    dei_names <- names(dcv_exp_info)
    channel_names <- dei_names[2:length(dei_names)]
    channels <- sapply(channel_names, function(channel_name) strsplit(channel_name, '_')[[1]][2])
    num_channels <- length(channels)
    names(channel_names) <- channels
    
    dcv_list <- get_full_calib_data(db_conn, dcv_exp_info)
    
    water_data <- dcv_list[['water']]
    well_nums <- colnames(water_data)
    num_wells <- length(well_nums)
    
    k_list_bydy <- lapply(channel_names, function(channel_name) dcv_list[[channel_name]] - water_data)
    names(k_list_bydy) <- channels
    
    dye_names <- paste('dye_', channels)
    
    k_inv_array <- array(
        0, 
        dim=c(num_channels, num_channels, num_wells), 
        dimnames=list(channels, dye_names, well_nums)
    )
    
    k_singular <- ''
    
    if (well_proc == 'mean') {
        k <- as.matrix(do.call(cbind, lapply(channels, function(channel) {
            k_data_1dye <- rowMeans(k_list_bydy[[channel]])
            k_data_1dye / sum(k_data_1dye)
        })))
        colnames(k) <- dye_names
        k_inv <- tryCatch(solve(k; suppress_warnings=true, error=err_e)
        if ('error' %in% class(k_inv)) {
            k_singular <- c('Well mean K matrix is singular. ')
        } else {
            for (well_num in well_nums) {
                k_inv_array[,,well_num] <- k_inv
            }
        }
        
    } else if (well_proc == 'dim3') {
        k <- k_inv_array
        k_singular_vec <- c()
        for (well_num in well_nums) {
            k_mtx <- as.matrix(do.call(cbind, lapply(channels, function(channel) {
                k_data_1dye <- k_list_bydy[[channel]][,well_num]
                k_data_1dye / sum(k_data_1dye)
            })))
            k[,,well_num] <- k_mtx
            k_inv <- tryCatch(solve(k_mtx; suppress_warnings=true), error=err_e)
            if ('error' %in% class(k_inv)) {
                k_singular_vec <- c(k_singular_vec, well_num)
            } else {
                k_inv_array[,,well_num] <- k_inv
            }
        }
        if (length(k_singular_vec) > 0) {
            k_singular <- sprintf(
                'Well-specific K matrix is singular for the following well(s): %s. ',
                paste(k_singular_vec, collapse=', ')
            )
        }
    }
    
    return(list('k'=k, 'k_inv_array'=k_inv_array, 'k_singular'=k_singular))
    }


# multi-channel deconvolution
deconv <- function(
    array2dcv, # dim1 must be channel, dim3 must be well; dim2 is cycle for amplification and temperature point for melting curve
    db_conn,
    calib_info
    ) {
    
    a2d_dim1 <- dim(array2dcv)[1]
    a2d_dim2 <- dim(array2dcv)[2]
    a2d_dim3 <- dim(array2dcv)[3]
    a2d_dimnames <- dimnames(array2dcv)
    
    # if data only has 1 cycle (amplification) or 1 temperature point (melt curve)
    if (is.na(a2d_dim3)) array2dcv <- array(c(array2dcv), 
                                            dim=c(a2d_dim1, 1, a2d_dim2), 
                                            dimnames=list(a2d_dimnames[[1]], '1', a2d_dimnames[[2]]))
    
    if (class(calib_info) == "numeric" || 
        sum(duplicated(sapply(calib_info, function(calib_ele) calib_ele[['step_id']]))) > 0
    ) {
        k_list_temp <- k_list
    } else {
        k_list_temp <- get_k(db_conn, calib_info, 'dim3')
    }
    
    k_inv_array = k_list_temp[['k_inv_array']]
    
    dcvd_by_dim2_well <- lapply(1:a2d_dim2, 
        function(dim2_i) do.call(cbind, lapply(1:a2d_dim3, 
            function(well_i) k_inv_array[,,well_i] %*% array2dcv[,dim2_i,well_i]))) # dim2 is cycle_num for amp and temperature for melt curve
    
    dcvd_array <- array(NA, dim(array2dcv))
    for (dim2_i in 1:dim(array2dcv)[2]) dcvd_array[,dim2_i,] <- dcvd_by_dim2_well[[dim2_i]]
    dimnames(dcvd_array) <- dimnames(array2dcv)
    
    # scale by channel to adjust for different fluorescence excitation strengths among dyes
    for (channel in dimnames(dcvd_array)[1]) {
        dcvd_array[channel,,] <- dcvd_array[channel,,] * scaling_factors_deconv[channel] }
    
    return(list('dcvd_array'=dcvd_array, 'k_list_temp'=k_list_temp))
    }

File: bioinformatics/julia/QpcrAnalysis/Essentials/deconv.jl
# color compensation / multi-channel deconvolution


# multi-channel deconvolution
function deconv(
    ary2dcv::AbstractArray, # dim1 is unit, which can be cycle (amplification), temperature point (melting curve), or step type (like "water", "channel_1", "channel_2" for calibration experiment); dim2 must be well, dim3 must be channel
    channels::AbstractVector, # must be the same length as 3rd dimension of `array2dcv`
    dcv_well_idc_wfluo::AbstractVector,

    # arguments needed k matrix needs to be computed
    db_conn::MySQL.MySQLHandle=db_conn_default, # `db_conn_default` is defined in "__init__.jl"
    calib_info::Union{Integer,OrderedDict}=calib_info_AIR,
    well_nums::AbstractVector=[];

    out_format::AbstractString="both" # "array", "dict", "both"
    )

    a2d_dim1, a2d_dim_well, a2d_dim_channel = size(ary2dcv)

    scaling_factor_dcv_vec = map(channels) do channel
        SCALING_FACTORS_deconv[channel]
    end

    dcvd_ary3 = similar(ary2dcv)

    k_dict = (isa(calib_info, Integer) || begin
        step_ids = map(ci_value -> ci_value["step_id"], values(calib_info))
        length_step_ids = length(step_ids)
        length_step_ids <= 2 || length(unique(step_ids)) < length_step_ids
    end) ? K_DICT : get_k(db_conn, calib_info, well_nums) # use default `well_proc` value

    k_inv_vec = k_dict["k_inv_vec"]

    for i1 in 1:a2d_dim1, i_well in 1:a2d_dim_well
        dcvd_ary3[i1, i_well, :] = *(
            k_inv_vec[dcv_well_idc_wfluo[i_well]],
            reshape(ary2dcv[i1, i_well, :], a2d_dim_channel)
        ) .* scaling_factor_dcv_vec
    end

    if out_format == "array"
        dcvd = (dcvd_ary3,)
    else
        dcvd_dict = OrderedDict(map(1:a2d_dim_channel) do channel_i
            channels[channel_i] => dcvd_ary3[:,:,channel_i]
        end) # do channel
        if out_format == "dict"
            dcvd = (dcvd_dict,)
        elseif out_format == "both"
            dcvd = (dcvd_ary3, dcvd_dict)
        else
            error("`out_format` must be \"array\", \"dict\" or \"both\".")
        end # if out_format == "dict"
    end # if out_format == "array"

    return (k_dict, dcvd...)

end # deconv


# function: get cross-over constant matrix k
function get_k(
    db_conn::MySQL.MySQLHandle, # MySQL database connection
    dcv_exp_info::OrderedDict, # OrderedDict("water"=OrderedDict(calibration_id=..., step_id=...), "channel_1"=OrderedDict(calibration_id=..., step_id=...),  "channel_2"=OrderedDict(calibration_id=...", step_id=...). # info on experiment(s) used to calculate matrix k
    well_nums::AbstractVector=[];
    well_proc::AbstractString="vec", # options: "mean", "vec".
    Float_T::DataType=Float32, # ensure compatibility with other OSs
    save_to::AbstractString="" # used: "k.jld"
    )

    dcv_exp_info = ensure_ci(db_conn, dcv_exp_info)

    calib_key_vec = get_ordered_keys(dcv_exp_info)
    cd_key_vec = calib_key_vec[2:end] # cd = channel of dye. "water" is index 1 per original order.

    dcv_data_dict = get_full_calib_data(db_conn, dcv_exp_info, well_nums)

    water_data, water_well_nums = dcv_data_dict["water"]
    num_wells = length(water_well_nums)

    k_dict_bydy = OrderedDict(map(cd_key_vec) do cd_key
        k_data_1dye, dcv_well_nums = dcv_data_dict[cd_key]
        return (cd_key, k_data_1dye .- water_data)
    end) # `dcv_well_nums` is not passed on because expected to be the same as `water_well_nums`, otherwise error will be raised by `get_full_calib_data`

    inv_note_pt1 = ""
    inv_note_pt2 = "K matrix is singular, using `pinv` instead of `inv` to compute inverse matrix of K. Deconvolution result may not be accurate. This may be caused by using the same or a similar set of solutions in the steps for different dyes. "

    if well_proc == "mean"
        k_s = hcat(
            map(cd_key_vec) do cd_key
                k_mean_vec_1dye = mean(k_dict_bydy[cd_key], 2)
                k_1dye = k_mean_vec_1dye / sum(k_mean_vec_1dye)
                return Array{Float_T}(k_1dye)
            end...) # do cd_key
        k_inv = try inv(k_s)
        catch err
            if isa(err, Base.LinAlg.SingularException)
                inv_note_pt1 = "Well mean"
                pinv(k_s)
            end # if isa(err,
        end # try
        k_inv_vec = fill(k_inv, num_wells)

    elseif well_proc == "vec"
        singular_well_nums = Vector{Int}()
        k_s = fill(ones(1,1), num_wells)
        k_inv_vec = similar(k_s)
        for i in 1:num_wells
            k_mtx = hcat(map(cd_key_vec) do cd_key
                k_vec_1dye = k_dict_bydy[cd_key][:,i]
                k_1dye = k_vec_1dye / sum(k_vec_1dye)
                return Array{Float_T}(k_1dye)
            end...) # do cd_key
            k_s[i] = k_mtx
            # k_inv_vec[i] = inv(k_mtx)
            k_inv_vec[i] = try inv(k_mtx)
            catch err
                if isa(err, Union{Base.LinAlg.SingularException, Base.LinAlg.LAPACKException})
                    push!(singular_well_nums, water_well_nums[i])
                    pinv(k_mtx)
                else
                    throw(err)
                end # if isa(err
            end # try
        end # for
        if length(singular_well_nums) > 0
            inv_note_pt1 = "Well(s) $(join(singular_well_nums, ", "))"
        end # if length

    end # if well_proc

    inv_note = length(inv_note_pt1) > 0 ? "$inv_note_pt1: $inv_note_pt2" : ""

    if length(save_to) > 0
        save(save_to,
            "k_s", k_s,
            "k_inv_vec", k_inv_vec,
            "inv_note", inv_note
        )
    end

    return OrderedDict(
        "k_s"=>k_s,
        "k_inv_vec"=>k_inv_vec,
        "inv_note"=>inv_note
    )

end # get_k
